src/java/org/apache/cassandra/db/memtable/AbstractMemtableWithCommitlog.java

```diff
@@ -112,9 +112,10 @@ public CommitLogPosition getCommitLogLowerBound()
         return commitLogLowerBound.get();
     }
 
-    public CommitLogPosition getCommitLogUpperBound()
+    public LastCommitLogPosition getCommitLogUpperBound()
     {
-        return commitLogUpperBound.get();
+        assert commitLogUpperBound.get() instanceof LastCommitLogPosition : "Commit log upper bound has not been sealed yet? " + commitLogUpperBound.get();
```

Comment:
commitLogUpperBound does not have default initialization.  If it is not set it will be null, not an empty AtomicReference

Reply:
Just saying I don't see a change to how commitLogUpperBound gets assigned. So it can be null.

Reply:
it now gets assigned before the write barrier. So it can't be null when `writeBarrier != null`.
Do I miss anything?

Reply:
Not missing anything. But what about future uses of this getter? That maybe someone adds something that could happen in a race again? So I would add a null check to the assert as well?  🤷‍♂️

Reply:
Why not assert that the barrier has expired? That must also be true at flush time, and would more clearly indicate why it's not okay to take the bound at an arbitrary time -- and we can add it to the method's description.